nssdb/encrypt.go

```diff
@@ -0,0 +1,359 @@
+	if err != nil {
+		return nil, fmt.Errorf("decode cipher text: %w", err)
+	}
+	if len(rest) > 0 {
+		return nil, fmt.Errorf("after decoding cipher text %d bytes remain", len(rest))
+	}
+
+	return edi, nil
+}
+
+// encodeCipherText takes raw encrypted data and formats it for storage along
+// with the pbes2 keygen (pbkdf2) and encryption (e.g. aes256-cbc) params used
+func encodeCipherText(encryptedData []byte, keyGenParams, encryptParams algIDGetter) ([]byte, error) {
+	keyGenAlgID, err := keyGenParams.getAlgorithmIdentifier()
+	if err != nil {
+		return nil, fmt.Errorf("get pbes2 key gen algorithm identifier: %w", err)
+	}
+	encryptAlgID, err := encryptParams.getAlgorithmIdentifier()
+	if err != nil {
+		return nil, fmt.Errorf("get pbes2 encryption algorithm identifier: %w", err)
+	}
+
+	pkcs5PBES2 := &pbes2Params{
+		KeyGen:     *keyGenAlgID,
+		Encryption: *encryptAlgID,
+	}
+	pbes2AlgID, err := pkcs5PBES2.getAlgorithmIdentifier()
+	if err != nil {
+		return nil, err
+	}
+
+	edi := encryptedDataInfo{
+		AlgorithmIdentifier: *pbes2AlgID,
+		EncryptedData:       encryptedData,
+	}
+	b, err := asn1.Marshal(edi)
+	if err != nil {
+		return nil, fmt.Errorf("marshal encrypted data info: %w", err)
+	}
+
+	return b, nil
+}
+
+func newAES256CBC() (*aes256CBCParams, error) {
+	iv, err := randutil.Bytes(aes.BlockSize)
+	if err != nil {
+		return nil, err
+	}
+	return &aes256CBCParams{
+		InitializationVector: iv,
+	}, nil
+}
+
+func (p aes256CBCParams) encrypt(key, plaintext []byte) ([]byte, error) {
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+
+	enc := cipher.NewCBCEncrypter(block, p.InitializationVector)
+
+	plaintext = cbcPad(plaintext)
+	ciphertext := make([]byte, len(plaintext))
+	enc.CryptBlocks(ciphertext, plaintext)
+
+	return ciphertext, nil
+}
+
+// https://github.com/nss-dev/nss/blob/NSS_3_107_RTM/lib/softoken/padbuf.c#L17
+func cbcPad(plaintext []byte) []byte {
+	inLen := len(plaintext)
+
+	desLen := (inLen + aes.BlockSize) & ^(aes.BlockSize - 1)
+	desPadLen := uint8(desLen - inLen)
+
+	for i := inLen; i < desLen; i++ {
+		plaintext = append(plaintext, desPadLen)
+	}
+
+	return plaintext
+}
+
+func cbcUnpad(plaintext []byte) []byte {
+	if len(plaintext) == 0 {
+		return nil
+	}
+	padLen := plaintext[len(plaintext)-1]
+	unpaddedLen := len(plaintext) - int(padLen)
+	if unpaddedLen < 1 {
+		return nil
+	}
+	return plaintext[:unpaddedLen]
+}
+
+func (p aes256CBCParams) decrypt(key, ciphertext []byte) ([]byte, error) {
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+	dec := cipher.NewCBCDecrypter(block, p.InitializationVector)
+	plaintext := make([]byte, len(ciphertext))
+	dec.CryptBlocks(plaintext, ciphertext)
+
+	out := cbcUnpad(plaintext)
+	if out == nil {
+		fmt.Printf("plain: %d, out %d\n", len(plaintext), len(out))
+		return nil, errors.New("failed to decrypt")
+	}
+	return out, nil
+}
+
+func (db NSSDB) encrypt(plaintext []byte) ([]byte, error) {
```

Comment:
Here and other places you probably want a `*NSSDB`.

Reply:
2168975